src/H5FDsubfiling/H5FDioc.c

```diff
@@ -807,158 +749,131 @@ H5FD__ioc_open(const char *name, unsigned flags, hid_t fapl_id, haddr_t maxaddr)
 
         /* Don't allow IOC threads to be finalized until everyone gets here */
-        if (!mpi_finalized && (file_ptr->mpi_size > 1))
-            if (MPI_SUCCESS != (mpi_code = MPI_Barrier(file_ptr->comm)))
-                H5_SUBFILING_MPI_GOTO_ERROR(FAIL, "MPI_Barrier failed", mpi_code);
+        if (!mpi_finalized && file->mpi_size > 1)
+            if (MPI_SUCCESS != (mpi_code = MPI_Barrier(file->comm)))
+                HMPI_GOTO_ERROR(FAIL, "MPI_Barrier failed", mpi_code);
 
-        /* Only finalize IOC threads and close subfiles if this is
-         * the last file holding a reference to the context
+        /* Only finalize IOC threads and close subfiles if this is the last file
+         * holding a reference to the context
          */
         if (sf_context && sf_context->file_ref == 1) {
-            if (sf_context->topology->rank_is_ioc && sf_context->threads_inited) {
-                if (finalize_ioc_threads(sf_context) < 0)
-                    /* Note that closing of subfiles is collective */
-                    H5_SUBFILING_DONE_ERROR(H5E_VFL, H5E_CANTCLOSEFILE, FAIL,
-                                            "unable to finalize IOC threads");
-            }
+            if (sf_context->topology->rank_is_ioc && sf_context->threads_inited)
+                if (H5FD__ioc_finalize_threads(sf_context) < 0)
+                    /* Fall through: closing of subfiles is collective */
+                    HDONE_ERROR(H5E_VFL, H5E_CANTCLOSEFILE, FAIL, "unable to finalize IOC threads");
 
-            if (H5_close_subfiles(file_ptr->context_id, file_ptr->comm) < 0)
-                H5_SUBFILING_GOTO_ERROR(H5E_VFL, H5E_CANTCLOSEFILE, FAIL,
-                                        "unable to close subfiling file(s)");
+            if (H5FD__subfiling_close_subfiles(file->context_id, file->comm) < 0)
+                HGOTO_ERROR(H5E_VFL, H5E_CANTCLOSEFILE, FAIL, "unable to close subfiling file(s)");
         }
-
-        file_ptr->context_id = -1;
     }
 
     if (!mpi_finalized) {
-        if (H5_mpi_comm_free(&file_ptr->comm) < 0)
-            H5_SUBFILING_GOTO_ERROR(H5E_VFL, H5E_CANTFREE, FAIL, "unable to free MPI Communicator");
-        if (H5_mpi_info_free(&file_ptr->info) < 0)
-            H5_SUBFILING_GOTO_ERROR(H5E_VFL, H5E_CANTFREE, FAIL, "unable to free MPI Info object");
+        if (H5_mpi_comm_free(&file->comm) < 0)
+            HGOTO_ERROR(H5E_VFL, H5E_CANTFREE, FAIL, "unable to free MPI Communicator");
+        if (H5_mpi_info_free(&file->info) < 0)
+            HGOTO_ERROR(H5E_VFL, H5E_CANTFREE, FAIL, "unable to free MPI Info object");
     }
 
-done:
-    free(file_ptr->file_path);
-    file_ptr->file_path = NULL;
-
-    H5MM_free(file_ptr->file_dir);
-    file_ptr->file_dir = NULL;
+    H5MM_free(file->file_path);
+    H5MM_free(file->file_dir);
 
     /* Release the file info */
-    file_ptr = H5FL_FREE(H5FD_ioc_t, file_ptr);
+    H5FL_FREE(H5FD_ioc_t, file);
 
-    H5_SUBFILING_FUNC_LEAVE;
+done:
```

Comment:
Why switch to skipping past the freeing of memory on error here? Just seems to leak memory for no reason. I'm also not a big fan of removing the

```
file_ptr->context_id = -1;
file_ptr->file_path = NULL;
file_ptr->file_dir = NULL;
```

statements since I like to be sure I leave memory in a sane state when free lists are involved, but it's not critical.